### PR TITLE
FSDP2 Support

### PR DESCRIPTION
**Pull request by intelava**
## Add FSDP2 support                                                                                                                                                     
                                                                                                                                                                           
  Added FSDP2 support to ignite using the composable `fully_shard` API.                                                                                                           
                                                                                                                                                                           
  ### What's new  
                                                                                                                                                                           
  auto_model now has a use_fsdp parameter. When True in a distributed context it                                                                                           
  applies fully_shard to the model instead of DDP. You can pass extra kwargs like
  mp_policy and offload_policy directly. Combining use_fsdp=True with sync_bn=True                                                                                         
  raises a ValueError right away since they don't work together.
                                                                                                                                                                           
  The checkpoint handler now recognizes FSDP2 models and uses get_model_state_dict
  and set_model_state_dict with full_state_dict=True so parameter gathering across                                                                                         
  ranks is handled automatically.                                                                                                                      
                  
  ### Tests                                                                                                                                                                
                  
  Added test_fsdp_smoke.py with 23 single process tests covering the flag                                                                                                  
  behavior, checkpoint imports and edge cases. test_fsdp_distributed.py is a
  standalone multi process script that spawns 2 workers with gloo and tests                                                                                                
  wrapping, forward pass and checkpoint save/load end to end.
                                                                                                                                                                           
  ---                                                                                                     
If you have any further questions or suggestions, I would be happy to help!

### CODE REVIEW

File: ignite/distributed/auto.py [hunk 1]
```diff
@@ -204,23 +243,33 @@ def auto_model(model: nn.Module, sync_bn: bool = False, **kwargs: Any) -> nn.Mod
...
-                lrank = idist.get_local_rank()
-                logger.info(f"Apply torch DistributedDataParallel on model, device id: {lrank}")
-                kwargs["device_ids"] = [
-                    lrank,
-                ]
+            if use_fsdp:
+                if not HAVE_FSDP2:
+                    raise RuntimeError(
+                        "fully_shard (FSDP2) is not available. Please upgrade to PyTorch >= 2.0."
+                    )
```
Comment (vfdev-5) [inline ignite/distributed/auto.py:256]: Minimal version of pytorch when fsdp2 appeared is 2.6

File: ignite/handlers/checkpoint.py [hunk 2]
```diff
@@ -403,6 +413,7 @@ def _get_filename_pattern(self, global_step: int | None) -> str:
                 with_score=self.score_function is not None,
                 with_score_name=self.score_name is not None,
                 with_global_step=global_step is not None,
+                with_ext=bool(self.ext),
```
Comment (vfdev-5) [inline ignite/handlers/checkpoint.py]: Let's rename this param as `as_folder`

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR @intelava !
I left few initial comments, and I'll review it in details a bit later

File: tests/ignite/test_fsdp_distributed.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+from __future__ import annotations
+
+import os
+import sys
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+
+sys.path.insert(0, "/pfss/mlde/workspaces/mlde_wsp_MazaheriA/tk27ryru/ignite")
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply tests/ignite/test_fsdp_distributed.py]: @intelava let's remove all local paths like this one
